irods/exception.py

```diff
@@ -69,6 +69,17 @@ class MultipleResultsFound(QueryException):
     pass
 
 
+class NotImplementedInIRODSServer(RuntimeError):
+    def __init__(self, feature_description, required_iRODS_version = ()):
+        super(NotImplementedInIRODSServer,self).__init__(feature_description + ': Not supported in connected iRODS server.')
+        self.required_iRODS_version = required_iRODS_version
+    def __str__(self):
+        return repr(self)
+    def __repr__(self):
+        nv = self.required_iRODS_version
+        return '{}{}'.format(self.args, ' [requires iRODS version: {nv}]'.format(**locals()) if nv else '')
```

Comment:
Maybe the `nv` should be "unknown" or some kind of error message instead of `''`?

Reply:
I kind of liked the idea of being able to omit the [ ] part totally if an iRODS version isn't handed to the constructor, but I don't mind changing toward consensus if "[Unknown]" seems better.

Reply:
or [required iRODS version unspecified]

Reply:
What does `required_iRODS_version` mean?
What is it used for?
Is the result of `__repr__` something that users can parse?
Is it meant to be parsed?

Reply:
Oh, I misunderstood the syntax... again... If `nv` evaluates to `False`, you will get an entirely empty string, not an empty string at the end of the "requires iRODS version" bit. In other words...
```python
>>> nv = None
>>> ' [requires iRODS version: {nv}]'.format(**locals()) if nv else ''
''
>>> nv = (4,3,2)
>>> ' [requires iRODS version: {nv}]'.format(**locals()) if nv else ''
' [requires iRODS version: (4, 3, 2)]'
```
So, you only see "requires iRODS version" if a version has been specified. Otherwise, it just shows the args.

I think this is okay because this makes specifying the required iRODS version an option of `NotImplementedInIRODSServer` but we could determine by some other means unrelated to version that `NotImplementedInIRODSServer` should be raised and we would not want to mention the version at all. Does that seem right, @d-w-moore?

Reply:
Yes, exactly, re the collapsing.
And yes, in the future other keyword arguments could be added to indicate other types of failure than version number. Or a lambda (does_server_support(session = sess) or so)plus a message template if one were ambitious, I guess....

Reply:
Should we resolve?

Reply:
I think @korydraughn had some questions, but I'm good on my end.

Reply:
Ignore my questions if @alanking's have been answered.

Reply:
> What does `required_iRODS_version` mean? What is it used for? Is the result of `__repr__` something that users can parse? Is it meant to be parsed?

`required_iRODS_version` is the minimum server version that will capably support the feature in question for use with PRC.
The __repr__ result may be parsed but the principle value is probably more that you might want to log such an exception and it's good to have the exception class reflected in that string.

Reply:
Got it. Thanks.